### PR TITLE
Added setup script for using CoaL on linux with KVM and libvirt.

### PR DESCRIPTION
**Pull request by siepkes**
I've created a script to setup CoaL on Linux with KVM and libvirt. I've tried to keep everything in sync with the CoaL docs as much as possible. The script is also pretty resilient and can be rerun to fix broken setups.

In the beginning I tried using the original CoaL VMDK's directly but I ran in to performance problems. In the end I configured libvirt to use the virtio-blk data plane. Using the virtio-blk data plane was the only way I got acceptable performance. The performance is now actually quite good! I ran the SDC setup and the setup completed in 779 seconds on my 4 year old Intel i7-3770K. I've no idea what the specs are of the machine on which the VMWare screenshots are taken but the screenshots show it took that machine 855 seconds. To use the virtio-blk data plane does however require the original CoaL images to be converted to raw format. Instructions are echo'ed by the script.

I tried using this configuration on Btrfs (my desktop was originally installed with Btrfs. Hey! I felt adventurous! ;-) but the performance was abysmal; Basically unusable, a lot of setup stuff would just timeout. I reinstalled my machine with XFS and everything was fine.

I have not tested this extensively but after running the setup (which showed no errors) I can SSH in to the headnode and everything seems to be running smoothly. I've also completed the post setup configuration (external NIC, cloudapi, etc.) and could open the SDC operations portal in my browser, browse around, etc.

```console
[root@headnode (coal-1) ~]# sdc-healthcheck 
ZONE                                 STATE           AGENT               STATUS
global                               running         -                   online
assets                               running         -                   online
sapi                                 running         -                   online
binder                               running         -                   online
manatee                              running         -                   online
moray                                running         -                   online
amonredis                            running         -                   online
redis                                running         -                   online
ufds                                 running         -                   online
workflow                             running         -                   online
amon                                 running         -                   online
sdc                                  running         -                   online
papi                                 running         -                   online
napi                                 running         -                   online
rabbitmq                             running         -                   online
imgapi                               running         -                   online
cnapi                                running         -                   online
dhcpd                                running         -                   online
fwapi                                running         -                   online
vmapi                                running         -                   online
ca                                   running         -                   online
mahi                                 running         -                   online
adminui                              running         -                   online
global                               running         ur                  online
global                               running         smartlogin          online
```

### CODE REVIEW

File: tools/coal-linux-kvm-setup [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+EXTERNAL_HOST_IP=10.88.88.2
+
+# Checks if a network exists and is active. Destroys and undefines it as needed.
+#
+# Argument 1: Name of the virsh network.
+#
+function delete_virsh_net() {
+    VIRSH_NET_NAME=$1
+
+    # GNU grep always exits with a non zero exit code if it didn't find the parttern.
```
Comment (fire) [inline tools/coal-linux-kvm-setup]: Typo.

**Comment by lloydde** [comment]
Very cool. I'd use this.

**Comment by fire** [comment]
Currently installing it on an AMD desktop with Ubuntu 16.04 (Pre-release) on a zfs drive.

I had to modify the script:
* unset kvm settings
* Use write back and thread mode for ZFS on ZFS (It should be possible to use a zfs device)
* Remove the virtio cmd line settings (Not sure what they did)

What performance tests can I use to see if there's a performance problem?

**Comment by siepkes** [comment]
Thanks for giving the script a spin! 

The performance problems I encountered were basically encountering abysmal performance. The whole thing was unusable. But as a "benchmark" I am interested to know how long it took for the final step to complete. It says something like: setup took X seconds right before it says setup complete. It is the step in which all headnode zones get created.

The qemu cli options enable the virtio data plane which results in beter IO performance. It's pretty new so it could be that ubuntu doesn't support them yet.

When using AMD you need to change the nested vmx setting too. Currently the script assumes you have an Intel CPU.

**Comment by fire** [comment]
I recall it was around 2500s, but there was another virtual machine running (Windows).

**Comment by fire** [comment]
Do you know how I would be able to have an external network on the NATed networking?

**Comment by siepkes** [comment]
2500 seems okay but somewhat high. I ended up with 755 but wasn't running an other VM.

For me on Fedora 23 my external network was immediately connected to the Internet. Maybe in Ubuntu it's needs additional configuration?

**Comment by fire** [comment]
What I want is to have coal use an external dhcp server to give vms ips.

**Comment by siepkes** [comment]
IP assignment of VM's is outside of the scope of the script; That's part of what SDC / Triton does. 

I also doubt you can use an external DHCP server for that. IP assignment of VM's is tightly integrated into SDC and as stated in the SDC design docs: "SmartDataCenter is very opinionated about how to architect a cloud.". SDC is best of in it's own network segment where it handles things the way SDC handles things (like IP assignment). Tightly integrating SDC in an existing infrastructure is not going to work without a lot of work.

**Comment by trentm** [comment]
@siepkes The "tools/" dir is for bits that get added as tools an the PATH in the headnode GZ. Please move your script to the "bin/" dir. I haven't tested this at all, but I'm cool adding this.

@joshwilsdon Any thoughts?

**Comment by trentm** [comment]
@siepkes Oh, wait, I was looking at the tools/ dir in sdc-headnode.git. Wrong repo :)

**Comment by trentm** [comment]
The tools/ dir sounds fine.

**Comment by trentm** [comment]
Merged in commit 113baf0
Thanks!

**Pull request closed without merging**